File: ECommerceApp/ServiceManager.m
//
//  ServiceManager.m
//  ECommerceApp
//
//  Created by Prateek Raj on 10/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

#import "ServiceManager.h"

static ServiceManager *sharedInstance = nil;

@implementation ServiceManager

+ (ServiceManager *) sharedInstance {
    
    @synchronized([ServiceManager class]) {
        
        if (!sharedInstance)            
            sharedInstance = [[self alloc] init];

        return sharedInstance;
    }
    return nil;
}

- (void)getDataFromServer {
    
    if (![DATAMANAGER checkIfDataAlreadyExists]) {
        
        NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:[NSURL URLWithString:@"https://stark-spire-93433.herokuapp.com/json"]
                                                               cachePolicy:NSURLRequestUseProtocolCachePolicy
                                                           timeoutInterval:10.0];
        [request setHTTPMethod:@"GET"];
        
        NSURLSession *session = [NSURLSession sharedSession];
        NSURLSessionDataTask *dataTask = [session dataTaskWithRequest:request
                                                    completionHandler:^(NSData *data, NSURLResponse *response, NSError *error) {
                                                        
                                                        if (error) {
                                                            
                                                            NSLog(@"%@", error);
                                                        } else {
                                                            
                                                            NSJSONSerialization *responseJson = [NSJSONSerialization JSONObjectWithData:data options:NSJSONReadingMutableLeaves error:nil];
                                                            
                                                            [DATAMANAGER removeAllExistingData];
                                                            
                                                            [DATAMANAGER pushDataToCoreData:(NSDictionary *)responseJson];
                                                        }
                                                    }];
        [dataTask resume];
    }
}







@end

File: ECommerceApp/ServiceManager.swift
//
//  ServiceManager.swift
//  ECommerceApp
//
//  Created by Prateek Raj on 28/07/18.
//  Copyright © 2018 Prateek Raj. All rights reserved.
//

import Foundation

class ServiceManager {
    public fileprivate(set) static var sharedInstance = ServiceManager()
    
    func getDataFromServer() {
        
        guard let url = URL(string: "https://stark-spire-93433.herokuapp.com/json") else { return }
        var request = URLRequest(url: url, cachePolicy: .useProtocolCachePolicy, timeoutInterval: 10)
        request.httpMethod = "GET"
        let urlSession = URLSession.shared
        let task = urlSession.dataTask(with: request, completionHandler: { (data: Data?, response: URLResponse?, error: Error?) -> Void in
            if let error = error {
                print(error)
            } else {
                guard let data = data else { return }
                guard let jsonResponse = try? JSONSerialization.jsonObject(with: data, options: .mutableLeaves), let dataReord = jsonResponse as? DataRecord else { return }
                DataManager.sharedInstance.removeAllExistingData()
                DataManager.sharedInstance.pushDataToCoreData(dataReord)
                
            }
        })
        task.resume()
    }
}
